### ISSUE TITLE
How do I replace UI graphics in specific songs?

### ISSUE THREAD

**Issue by dombomb64** [ShadowMario/FNF-PsychEngine#9211]
### Describe your problem here.

In my mod, I want to replace the countdown and rating graphics with my own versions in specific songs (meaning no replacing existing files). However, the only way I can think of accomplishing this is by using the `loadGraphic()` function in the source.

I don't actually know if that's possible, but my mod is completely Lua-based, so it isn't viable anyway because you can't run methods from the source in Lua scripts. Could there be a variable in the sprites that I could change to change the graphic?

P.S. Before posting this, I looked in the HaxeFlixel API one more time, and in `FlxGraphic`, I found the `BitmapData` variable `bitmap`. Would it be viable to make a dummy Lua sprite and copy over its `<sprite>.graphic.bitmap` variable to the sprites from the source?

TL;DR, I want to change the style of the countdown and rating graphics without replacing assets, but I don't know how.

### Are you modding a build from source or with Lua?

Lua

### What is your build target?

Windows x64

### Did you edit anything in this build? If so, mention or summarize your changes.

No.

**Comment by 0-skid** [comment]
idk about the countdown but what do you mean by rating graphics?

**Comment by dombomb64** [comment]
I meant the combo counter and timing graphics e.g. "Good" or "Sick".

**Comment by frantastic24** [comment]
```lua
function onCreatePost()
	setProperty('ratingsData[0].image', 'image') --change Sick image, change array for other ratings
end

function onCountdownTick(counter)
	loadGraphic('countdownReady', 'image')-- countdownReady/Set/Go
end
```

**Comment by dombomb64** [comment]
Thanks, but the only thing that didn't work was the graphic for "Go". I've checked the file name to make sure it's correct, and it is.

Also, do you know how to change the graphics for the "Combo" sprite and the combo numbers? I've looked through the code, and it looks like the rating sprites, "Combo" sprites, and number sprites all get inserted into the array `strumLineNotes` for some reason. Is that correct?

**Comment by frantastic24** [comment]
> but the only thing that didn't work was the graphic for "Go"

i can confirm it works, the problem might be that `onCountdownTick()` tends to fail sometimes? it happened to me when i had another script using it. i'm not sure why it happens but i recommend you check with `debugPrint()` that it is getting called and that the entire code is executing

> do you know how to change the graphics for the "Combo" sprite and the combo numbers?

i don't think you can, they are generated each time and there's no variable tied to them you can change to do this

> sprites all get inserted into the FlxTypedGroup strumLineNotes for some reason

for layering, wouldn't want to have them be _on top_ of the strums

**Comment by dombomb64** [comment]
> > but the only thing that didn't work was the graphic for "Go"
> 
> i can confirm it works, the problem might be that `onCountdownTick()` tends to fail sometimes? it happened to me when i had another script using it. i'm not sure why it happens but i recommend you check with `debugPrint()` that it is getting called and that the entire code is executing

I tried checking with `debugPrint()` and sometimes it would print twice, but usually it printed three times.

> > do you know how to change the graphics for the "Combo" sprite and the combo numbers?
> 
> i don't think you can, they are generated each time and there's no variable tied to them you can change to do this

Wouldn't it be possible to run `loadGraphic()` on everything in `strumLineNotes` that has a specific image?

> > sprites all get inserted into the FlxTypedGroup strumLineNotes for some reason
> 
> for layering, wouldn't want to have them be _on top_ of the strums

?

**Comment by frantastic24** [comment]
> I tried checking with debugPrint(), and sometimes it would print twice, but usually it printed three times.

the function is called 5 times so there could be an issue

> Wouldn't it be possible to run loadGraphic() on everything in strumLineNotes that has a specific image?

strumLineNotes has nothing to do with this, like i said that is there just for layering

**Comment by dombomb64** [comment]
> the function is called 5 times so there could be an issue

I tried changing `onCountdownTick()` to `onUpdate()`, but now, only the "Ready" graphic is replaced despite the condition I nested the `loadGraphic()` calls in, `if curStep < 1`, being met. The `debugPrint()` call is working, so I don't know why the `loadGraphic()` calls aren't. I would think that it might not like being called so frequently, but I just tried lowering the rate at which it does so, and it still doesn't work.

> `strumLineNotes` has nothing to do with this, like i said that is there just for layering

Okay, I'm going to try to make my own Lua combo sprites to replace them.

**Comment by frantastic24** [comment]
well if you are loading a graphic each frame yeah it's not going to like it, you could probably use timers as a work around

**Comment by dombomb64** [comment]
I just tried using timers, and it still only loads the first one. Maybe I'll just make my own sprites.

**Comment by frantastic24** [comment]
Just to be sure, do you mind posting the whole script? You can send previous versions too to make sure there isn't something wrong with those. And also did you check if you had any other script using `onCountdownTick()` like I mentioned?
Put them in a code box here preferably

**Comment by dombomb64** [comment]
nutshellUi.lua
```
local isNutshell = false;

function onCreatePost()
	for i = 0, getProperty('unspawnNotes.length') - 1 do
		if getPropertyFromGroup('unspawnNotes', i, 'noteSplashTexture') == 'noteSplashesNutshell' or getPropertyFromGroup('unspawnNotes', i, 'noteSplashTexture') == 'hurtNoteSplashesNutshell' or getPropertyFromGroup('unspawnNotes', i, 'noteSplashTexture') == 'crucialNoteSplashesNutshell' then
			isNutshell = true;
		end
	end

	if isNutshell == false then
		--debugPrint('test');
		return;
	end

	precacheImage('nutshellUi/sick');
	precacheImage('nutshellUi/good');
	precacheImage('nutshellUi/bad');
	precacheImage('nutshellUi/shit');
	precacheImage('nutshellUi/ready');
	precacheImage('nutshellUi/set');
	precacheImage('nutshellUi/go');
	precacheImage('nutshellUi/combo');
	precacheImage('nutshellUi/num0');
	precacheImage('nutshellUi/num1');
	precacheImage('nutshellUi/num2');
	precacheImage('nutshellUi/num3');
	precacheImage('nutshellUi/num4');
	precacheImage('nutshellUi/num5');
	precacheImage('nutshellUi/num6');
	precacheImage('nutshellUi/num7');
	precacheImage('nutshellUi/num8');
	precacheImage('nutshellUi/num9');

	setProperty('ratingsData[0].image', 'nutshellUi/sick');
	setProperty('ratingsData[1].image', 'nutshellUi/good');
	setProperty('ratingsData[2].image', 'nutshellUi/bad');
	setProperty('ratingsData[3].image', 'nutshellUi/shit');

	--runTimer('loadCountdownGraphics', 0.05, 100);
end

local loadGraphicCooldown = 0;

--[[function onUpdate(elapsed)
	if isNutshell == false then
		return;
	end

	--if curStep < 1 and loadGraphicCooldown == 0 then
	if curStep < 1 then
		loadGraphic('countdownReady', 'nutshellUi/ready');
		loadGraphic('countdownSet', 'nutshellUi/set');
		loadGraphic('countdownGo', 'nutshellUi/go');
		--loadGraphicCooldown = 10;
		--debugPrint(curStep);
	--else
		--loadGraphicCooldown = loadGraphicCooldown - 1;
	end
end]]--

--function onSongStart()
--end

function onTimerCompleted(tag, loops, loopsLeft)
	if tag == 'loadCountdownGraphics' then
		loadGraphic('countdownReady', 'nutshellUi/ready');
		loadGraphic('countdownSet', 'nutshellUi/set');
		loadGraphic('countdownGo', 'nutshellUi/go');
		--debugPrint('test');
	end
end
```

backupCharacters.lua (onCountdownTick function only)
```
function onCountdownTick(counter)
	if counter % 2 == 0 then
		objectPlayAnimation('boyfriend2', 'idle', true);
		objectPlayAnimation('dad2', 'idle', true);
	end
end
```

legs.lua (onCountdownTick function only)
```
function onCountdownTick(counter)
	if counter % 2 == 0 then
		objectPlayAnimation('kristenLegs', 'danceLeft', true);
		objectPlayAnimation('kristenLegsPlayer', 'danceLeft', true);
		objectPlayAnimation('adamLegs', 'danceLeft', true);
		objectPlayAnimation('adamLegsPlayer', 'danceLeft', true);
	elseif (counter + 1) % 2 == 0 then
		objectPlayAnimation('kristenLegs', 'danceRight', true);
		objectPlayAnimation('kristenLegsPlayer', 'danceRight', true);
		objectPlayAnimation('adamLegs', 'danceRight', true);
		objectPlayAnimation('adamLegsPlayer', 'danceRight', true );
	end
	objectPlayAnimation('adamStuckLegs', 'idle', true);
	objectPlayAnimation('adamStuckLegsPlayer', 'idle', true);
end
```

P.S. Just before posting this, I tried to combine all the `onCountdownTick` calls into on function, but it still didn't change the graphic for "Go".

**Comment by frantastic24** [comment]
That timer doesn't work, the idea is to have it run based on beat duration just like the actual countdown is timed
I don't see why fusing functions didn't work, assuming you commented all the other ones out. Does it still not work with another image?

**Comment by dombomb64** [comment]
Yes, I commented all the other ones out, and I just tried using another image to no avail.

**Comment by frantastic24** [comment]
I just remembered arrays can be edited now so I'm taking a shot in the dark
```lua
function onCountdownTick()
     setProperty('introAlts[0]', 'image path') --ready
     setProperty('introAlts[1]', 'image path') --set
     setProperty('introAlts[2]', 'image path') --go
end
```

**Comment by dombomb64** [comment]
It doesn't do anything...

**Comment by frantastic24** [comment]
okay yeah it was worth a shot though
my initial solution is working just fine for me so this is just something on your end i can't really help with

**Comment by dombomb64** [comment]
Here's the code, as promised.

```
local isNutshell = false;

local showCombo = false;
local showComboNum = true;
local comboIdCounter = 9; -- 0 to 9, so there will be at most 10 sets of combo graphics at once

function onCreatePost()
	showCombo = getProperty('showCombo');
	showComboNum = getProperty('showComboNum');

	for i = 0, getProperty('unspawnNotes.length') - 1 do
		if getPropertyFromGroup('unspawnNotes', i, 'noteSplashTexture') == 'noteSplashesNutshell' or getPropertyFromGroup('unspawnNotes', i, 'noteSplashTexture') == 'hurtNoteSplashesNutshell' or getPropertyFromGroup('unspawnNotes', i, 'noteSplashTexture') == 'crucialNoteSplashesNutshell' then
			isNutshell = true;
		end
	end

	if isNutshell == false then
		--debugPrint('test');
		return;
	end

	setProperty('showCombo', false);
	setProperty('showComboNum', false);

	precacheImage('nutshellUi/sick');
	precacheImage('nutshellUi/good');
	precacheImage('nutshellUi/bad');
	precacheImage('nutshellUi/shit');
	precacheImage('nutshellUi/ready');
	precacheImage('nutshellUi/set');
	precacheImage('nutshellUi/go');
	precacheImage('nutshellUi/combo');
	precacheImage('nutshellUi/num0');
	precacheImage('nutshellUi/num1');
	precacheImage('nutshellUi/num2');
	precacheImage('nutshellUi/num3');
	precacheImage('nutshellUi/num4');
	precacheImage('nutshellUi/num5');
	precacheImage('nutshellUi/num6');
	precacheImage('nutshellUi/num7');
	precacheImage('nutshellUi/num8');
	precacheImage('nutshellUi/num9');

	setProperty('ratingsData[0].image', 'nutshellUi/sick');
	setProperty('ratingsData[1].image', 'nutshellUi/good');
	setProperty('ratingsData[2].image', 'nutshellUi/bad');
	setProperty('ratingsData[3].image', 'nutshellUi/shit');

	runTimer('loadCountdownGraphics', crochet / 1000, 5);

	--[[setProperty('introAlts[0]', 'nutshellUi/ready');
	setProperty('introAlts[1]', 'nutshellUi/set');
	setProperty('introAlts[2]', 'nutshellUi/go');]]--
	--popUpScore();
	--runTimer('1comboNumFade0', crochet * 0.002, 1);
end

--local loadGraphicCooldown = 0;

--[[function onUpdate(elapsed)
	if isNutshell == false then
		return;
	end

	--if curStep < 1 and loadGraphicCooldown == 0 then
	if curStep < 1 then
		loadGraphic('countdownReady', 'nutshellUi/ready');
		loadGraphic('countdownSet', 'nutshellUi/set');
		loadGraphic('countdownGo', 'nutshellUi/go');
		--loadGraphicCooldown = 10;
		--debugPrint(curStep);
	--else
		--loadGraphicCooldown = loadGraphicCooldown - 1;
	end
end]]--

--function onSongStart()
--end

function onTimerCompleted(tag, loops, loopsLeft)
	if tag == 'loadCountdownGraphics' then
		if loopsLeft == 3 then
			loadGraphic('countdownReady', 'nutshellUi/ready');
			setProperty('countdownReady.antialiasing', false);
		elseif loopsLeft == 2 then
			loadGraphic('countdownSet', 'nutshellUi/set');
			setProperty('countdownSet.antialiasing', false);
		elseif loopsLeft == 1 then
			loadGraphic('countdownGo', 'nutshellUi/go');
			setProperty('countdownGo.antialiasing', false);
		end
		--debugPrint('test');
	elseif string.find(tag, 'comboNumFade') ~= nil then
		local digit = string.sub(tag, 1, 1);
		local id = string.sub(tag, -1);
		--debugPrint(digit .. ', ' .. id);
		doTweenAlpha(tag, digit .. 'comboNum' .. id, 0, 0.2, 'linear');
	elseif string.find(tag, 'comboSpriteFade') ~= nil then
		--local digit = string.sub(tag, 1, 1);
		local id = string.sub(tag, -1);
		--debugPrint(digit .. ', ' .. id);
		doTweenAlpha(tag, 'comboSprite' .. id, 0, 0.2, 'linear');
	end
end

function onTweenCompleted(tag)
	if string.find(tag, 'comboNumFade') ~= nil then
		local digit = string.sub(tag, 1, 1);
		local id = string.sub(tag, -1);
		removeLuaSprite(digit .. 'comboNum' .. id, true);
	elseif string.find(tag, 'comboSpriteFade') ~= nil then
		local id = string.sub(tag, -1);
		--debugPrint(id);
		removeLuaSprite('comboSprite' .. id, true);
	end
end

function goodNoteHit(id, direction, noteType, isSustainNote)
	--debugPrint(getPropertyFromGroup('notes', id, 'hitCausesMiss'));
	if not isSustainNote and not getPropertyFromGroup('notes', id, 'hitCausesMiss') then
		popUpScore();
	end
end

function popUpScore()
	-- Update ID Counter
	if comboIdCounter == 9 then
		comboIdCounter = 0;
	else
		comboIdCounter = comboIdCounter + 1;
	end
	-- "Combo" Text
	comboSpriteName = 'comboSprite' .. comboIdCounter;

	makeLuaText('comboText', tostring(getProperty('combo')), 0, screenWidth * 0.35, 0);
	setTextSize('comboText', 32);
	setProperty('comboText.y', getProperty('camHUD.height') / 2 - getProperty('comboText.height') / 2);

	makeLuaSprite(comboSpriteName, 'nutshellUi/combo', getProperty('comboText.x'), getProperty('comboText.y'));
	setObjectCamera(comboSpriteName, 'hud');
	--setProperty(comboSpriteName .. '.x', getProperty('comboText.x'));
	setProperty(comboSpriteName .. '.acceleration.y', getRandomInt(200, 300));
	setProperty(comboSpriteName .. '.velocity.y', -getRandomInt(140, 160));
	setProperty(comboSpriteName .. '.x', getProperty(comboSpriteName .. '.x') + getPropertyFromClass('ClientPrefs', 'comboOffset[0]'));
	setProperty(comboSpriteName .. '.y', getProperty(comboSpriteName .. '.y') - getPropertyFromClass('ClientPrefs', 'comboOffset[1]') + 60);
	if not hideHud and showCombo then
		addLuaSprite(comboSpriteName, true);
	end

	--local comboSpriteScale = math.floor(getProperty(comboSpriteName .. '.width') * 0.7);
	setGraphicSize(comboSpriteName, math.floor(getProperty(comboSpriteName .. '.width') * 0.7), math.floor(getProperty(comboSpriteName .. '.height') * 0.7));
	--setProperty(comboSpriteName .. '.scale.x', comboSpriteScale);
	--setProperty(comboSpriteName .. '.scale.y', comboSpriteScale);
	setProperty(comboSpriteName .. '.antialiasing', false);
	--debugPrint('test');
	--debugPrint(getProperty(comboSpriteName .. '.scale.x') .. ', ' .. getProperty(comboSpriteName .. '.scale.y'));

	local separatedScore = {};
	if getProperty('combo') >= 1000 then
		table.insert(separatedScore, math.floor(getProperty('combo') / 1000) % 10);
	end
	table.insert(separatedScore, math.floor(getProperty('combo') / 100) % 10);
	table.insert(separatedScore, math.floor(getProperty('combo') / 10) % 10);
	table.insert(separatedScore, getProperty('combo') % 10);
	--debugPrint(separatedScore);

	-- Combo Numbers
	for i, v in ipairs(separatedScore) do
		local numSpriteName = i .. 'comboNum' .. comboIdCounter;

		makeLuaSprite(numSpriteName, 'nutshellUi/num' .. v);
		setObjectCamera(numSpriteName, 'hud');
		setProperty(numSpriteName .. '.x', getProperty('comboText.x') + (43 * (i - 1)) - 90 + getPropertyFromClass('ClientPrefs', 'comboOffset[2]'));
		setProperty(numSpriteName .. '.y', getProperty('camHUD.height') / 2 - getProperty(numSpriteName .. '.height') / 2 + 80 - getPropertyFromClass('ClientPrefs', 'comboOffset[3]'));
		setGraphicSize(numSpriteName, math.floor(getProperty(numSpriteName .. '.width') * 0.5), math.floor(getProperty(numSpriteName .. '.height') * 0.5));
		setProperty(numSpriteName .. '.antialiasing', false);
		setProperty(numSpriteName .. '.acceleration.y', getRandomInt(200, 300));
		setProperty(numSpriteName .. '.velocity.y', -getRandomInt(140, 160));
		setProperty(numSpriteName .. '.velocity.x', getRandomInt(-5, 5));
		if not hideHud and showComboNum then
			addLuaSprite(numSpriteName, true);
		end
		runTimer(i .. 'comboNumFade' .. comboIdCounter, crochet * 0.002, 1);
		--doTweenAlpha(i .. 'comboNumFade' .. comboIdCounter, numSpriteName, 0, 0.2, 'linear');
	end
	runTimer('comboSpriteFade' .. comboIdCounter, crochet * 0.002, 1);
end
```

**Comment by dombomb64** [comment]
I figured it out, but I forgot to post my solution. Since `onCountdownTick` seems to be working again, I just used that instead of running a timer.

My code:

```
-- NOTE:
-- Feel free to use this script in your mods!
-- Just make sure to give credit to me, dombomb64.
-- (You can delete all the commented-out code, most of it is remnants from past attempts.)
-- Related Scripts: customBars, flipnoteUi, pixelArrows, pixelIcons (You can use those, too.)

local isNutshell = false;

local showCombo = false;
local showComboNum = true;
local comboIdCounter = 9; -- 0 to 9, so there will be at most 10 sets of combo graphics at once

function onCreatePost()
	showCombo = getProperty('showCombo');
	showComboNum = getProperty('showComboNum');

	for i = 0, getProperty('unspawnNotes.length') - 1 do
		if getPropertyFromGroup('unspawnNotes', i, 'noteSplashTexture') == 'noteSplashesNutshell' or getPropertyFromGroup('unspawnNotes', i, 'noteSplashTexture') == 'hurtNoteSplashesNutshell' or getPropertyFromGroup('unspawnNotes', i, 'noteSplashTexture') == 'crucialNoteSplashesNutshell' then
			isNutshell = true;
			--debugPrint('nutshell');
			break;
		end
	end

	if isNutshell == false then
		--debugPrint('test');
		return;
	end

	setProperty('showCombo', false);
	setProperty('showComboNum', false);

	precacheImage('nutshellUi/sick');
	precacheImage('nutshellUi/good');
	precacheImage('nutshellUi/bad');
	precacheImage('nutshellUi/shit');
	precacheImage('nutshellUi/ready');
	precacheImage('nutshellUi/set');
	precacheImage('nutshellUi/go');
	precacheImage('nutshellUi/combo');
	precacheImage('nutshellUi/num0');
	precacheImage('nutshellUi/num1');
	precacheImage('nutshellUi/num2');
	precacheImage('nutshellUi/num3');
	precacheImage('nutshellUi/num4');
	precacheImage('nutshellUi/num5');
	precacheImage('nutshellUi/num6');
	precacheImage('nutshellUi/num7');
	precacheImage('nutshellUi/num8');
	precacheImage('nutshellUi/num9');

	setProperty('ratingsData[0].image', 'nutshellUi/sick');
	setProperty('ratingsData[1].image', 'nutshellUi/good');
	setProperty('ratingsData[2].image', 'nutshellUi/bad');
	setProperty('ratingsData[3].image', 'nutshellUi/shit');

	--runTimer('loadCountdownGraphics', crochet / 1000, 5);

	--[[setProperty('introAlts[0]', 'nutshellUi/ready');
	setProperty('introAlts[1]', 'nutshellUi/set');
	setProperty('introAlts[2]', 'nutshellUi/go');]]--
	--popUpScore();
	--runTimer('1comboNumFade0', crochet * 0.002, 1);
end

--[[function onStartCountdown()
	runTimer('loadCountdownGraphicsNutshell', crochet / 1000, 5);
end]]--

--local loadGraphicCooldown = 0;

--[[function onUpdate(elapsed)
	if isNutshell == false then
		return;
	end

	--if curStep < 1 and loadGraphicCooldown == 0 then
	if curStep < 1 then
		loadGraphic('countdownReady', 'nutshellUi/ready');
		loadGraphic('countdownSet', 'nutshellUi/set');
		loadGraphic('countdownGo', 'nutshellUi/go');
		--loadGraphicCooldown = 10;
		--debugPrint(curStep);
	--else
		--loadGraphicCooldown = loadGraphicCooldown - 1;
	end
end]]--

--function onSongStart()
--end

function onCountdownTick(counter)
	--debugPrint('test');
	if isNutshell then
		if counter == 1 then
			--setProperty('countdownReady.offset.y', 5000);
			--runTimer('removeCountdownGraphics', 0, 1);
			--removeLuaSprite('countdownReady', true);
			loadGraphic('countdownReady', 'nutshellUi/ready');
			--setProperty('countdownReady.scale.x', 1);
			setProperty('countdownReady.antialiasing', false);
			--debugPrint('test');
			--makeLuaSprite('ready', 'nutshellUi/flipnote/ready', 0, 0);
			--setObjectCamera('ready', 'hud');
			--screenCenter('ready');
			--addLuaSprite('ready', true);
		elseif counter == 2 then
			--setProperty('countdownSet.offset.y', 5000);
			--runTimer('removeCountdownGraphics', 0, 1);
			--removeLuaSprite('countdownSet', true);
			loadGraphic('countdownSet', 'nutshellUi/set');
			setProperty('countdownSet.antialiasing', false);
		elseif counter == 3 then
			--setProperty('countdownGo.offset.y', 5000);
			--runTimer('removeCountdownGraphics', 0, 1);
			--removeLuaSprite('countdownGo', true);
			loadGraphic('countdownGo', 'nutshellUi/go');
			setProperty('countdownGo.antialiasing', false);
		end
	end
end

function onTimerCompleted(tag, loops, loopsLeft)
	if tag == 'loadCountdownGraphicsNutshell' then
		if loopsLeft == 3 then
			loadGraphic('countdownReady', 'nutshellUi/ready');
			--setProperty('countdownReady.scale.x', 1);
			setProperty('countdownReady.antialiasing', false);
			--debugPrint('test');
		elseif loopsLeft == 2 then
			loadGraphic('countdownSet', 'nutshellUi/set');
			setProperty('countdownSet.antialiasing', false);
		elseif loopsLeft == 1 then
			loadGraphic('countdownGo', 'nutshellUi/go');
			setProperty('countdownGo.antialiasing', false);
		end
		--debugPrint('test');
	elseif string.find(tag, 'comboNumFade') ~= nil then
		local digit = string.sub(tag, 1, 1);
		local id = string.sub(tag, -1);
		--debugPrint(digit .. ', ' .. id);
		doTweenAlpha(tag, digit .. 'comboNum' .. id, 0, 0.2, 'linear');
	elseif string.find(tag, 'comboSpriteFade') ~= nil then
		--local digit = string.sub(tag, 1, 1);
		local id = string.sub(tag, -1);
		--debugPrint(digit .. ', ' .. id);
		doTweenAlpha(tag, 'comboSprite' .. id, 0, 0.2, 'linear');
	end
end

function onTweenCompleted(tag)
	if string.find(tag, 'comboNumFade') ~= nil then
		local digit = string.sub(tag, 1, 1);
		local id = string.sub(tag, -1);
		removeLuaSprite(digit .. 'comboNum' .. id, true);
	elseif string.find(tag, 'comboSpriteFade') ~= nil then
		local id = string.sub(tag, -1);
		--debugPrint(id);
		removeLuaSprite('comboSprite' .. id, true);
	end
end

function goodNoteHit(id, direction, noteType, isSustainNote)
	--debugPrint(getPropertyFromGroup('notes', id, 'hitCausesMiss'));
	if not isSustainNote and not getPropertyFromGroup('notes', id, 'hitCausesMiss') and isNutshell then
		popUpScore();
	end
end

function popUpScore()
	-- Update ID Counter
	if comboIdCounter == 9 then
		comboIdCounter = 0;
	else
		comboIdCounter = comboIdCounter + 1;
	end
	-- "Combo" Text
	comboSpriteName = 'comboSprite' .. comboIdCounter;

	makeLuaText('comboText', tostring(getProperty('combo')), 0, screenWidth * 0.35, 0);
	setTextSize('comboText', 32);
	setProperty('comboText.y', getProperty('camHUD.height') / 2 - getProperty('comboText.height') / 2);

	makeLuaSprite(comboSpriteName, 'nutshellUi/combo', getProperty('comboText.x'), getProperty('comboText.y'));
	setObjectCamera(comboSpriteName, 'hud');
	--setProperty(comboSpriteName .. '.x', getProperty('comboText.x'));
	setProperty(comboSpriteName .. '.acceleration.y', getRandomInt(200, 300));
	setProperty(comboSpriteName .. '.velocity.y', -getRandomInt(140, 160));
	setProperty(comboSpriteName .. '.x', getProperty(comboSpriteName .. '.x') + getPropertyFromClass('ClientPrefs', 'comboOffset[0]'));
	setProperty(comboSpriteName .. '.y', getProperty(comboSpriteName .. '.y') - getPropertyFromClass('ClientPrefs', 'comboOffset[1]') + 60);
	if not hideHud and showCombo then
		addLuaSprite(comboSpriteName, false);
	end

	--local comboSpriteScale = math.floor(getProperty(comboSpriteName .. '.width') * 0.7);
	setGraphicSize(comboSpriteName, math.floor(getProperty(comboSpriteName .. '.width') * 0.7), math.floor(getProperty(comboSpriteName .. '.height') * 0.7));
	--setProperty(comboSpriteName .. '.scale.x', comboSpriteScale);
	--setProperty(comboSpriteName .. '.scale.y', comboSpriteScale);
	setProperty(comboSpriteName .. '.antialiasing', false);
	--debugPrint('test');
	--debugPrint(getProperty(comboSpriteName .. '.scale.x') .. ', ' .. getProperty(comboSpriteName .. '.scale.y'));

	local separatedScore = {};
	if getProperty('combo') >= 1000 then
		table.insert(separatedScore, math.floor(getProperty('combo') / 1000) % 10);
	end
	table.insert(separatedScore, math.floor(getProperty('combo') / 100) % 10);
	table.insert(separatedScore, math.floor(getProperty('combo') / 10) % 10);
	table.insert(separatedScore, getProperty('combo') % 10);
	--debugPrint(separatedScore);

	-- Combo Numbers
	for i, v in ipairs(separatedScore) do
		local numSpriteName = i .. 'comboNum' .. comboIdCounter;

		makeLuaSprite(numSpriteName, 'nutshellUi/num' .. v);
		setObjectCamera(numSpriteName, 'hud');
		setProperty(numSpriteName .. '.x', getProperty('comboText.x') + (43 * (i - 1)) - 90 + getPropertyFromClass('ClientPrefs', 'comboOffset[2]'));
		setProperty(numSpriteName .. '.y', getProperty('camHUD.height') / 2 - getProperty(numSpriteName .. '.height') / 2 + 80 - getPropertyFromClass('ClientPrefs', 'comboOffset[3]'));
		setGraphicSize(numSpriteName, math.floor(getProperty(numSpriteName .. '.width') * 0.5), math.floor(getProperty(numSpriteName .. '.height') * 0.5));
		setProperty(numSpriteName .. '.antialiasing', false);
		setProperty(numSpriteName .. '.acceleration.y', getRandomInt(200, 300));
		setProperty(numSpriteName .. '.velocity.y', -getRandomInt(140, 160));
		setProperty(numSpriteName .. '.velocity.x', getRandomInt(-5, 5));
		if not hideHud and showComboNum then
			addLuaSprite(numSpriteName, false);
		end
		runTimer(i .. 'comboNumFade' .. comboIdCounter, crochet * 0.002, 1);
		--doTweenAlpha(i .. 'comboNumFade' .. comboIdCounter, numSpriteName, 0, 0.2, 'linear');
	end
	runTimer('comboSpriteFade' .. comboIdCounter, crochet * 0.002, 1);
end
```
